src/vector/geometry.rs

```diff
@@ -480,9 +514,45 @@ impl Debug for GeometryRef<'_> {
     }
 }
 
+/// Invocation options for [`Geometry::make_valid_ex`].
+#[derive(Debug, Clone, Default)]
+pub enum MakeValidOpts {
+    /// Combines all rings into a set of node-ed lines and then extracts valid polygons from that "linework".
+    #[default]
+    Linework,
+    /// First makes all rings valid, then merges shells and subtracts holes from shells to generate valid result.
+    ///
+    /// Assumes holes and shells are correctly categorized.
+    Structure {
+        /// If `false`, collapses are converted to empty geometries.
+        /// If `true`, collapses are converted to a valid geometry of lower dimension.
+        keep_collapsed: bool,
+    },
+}
```

Comment:
Purpose here is to try to encode possible options in a way that only expresses valid combination of options. I'd like to think our attempting to provide these ergonomics are in-scope for this library, but I could see arguments against it too.

w.r.t. the opening question: should we just have a single `make_valid` that always requires this (e.g. `geom.make_valid(MakeValidOpts::default())`) or is it better have two separate methods (as this implementation has)?

Reply:
Yeah, but the valid combinations today might be insufficient tomorrow. We could either allow passing a `CSLStringList`, or add a `Custom` variant here -- but then what happens when GDAL adds another unrelated option?

Reply:
How about this approach?

https://github.com/georust/gdal/pull/356#discussion_r1059668374

It's in times like these I wish Rust had default arguments...